query/services/ingestion.py

```diff
@@ -0,0 +1,199 @@
+import logging
+import math
+from datetime import datetime, timezone
+from typing import Dict
+
+from asyncpg import Connection
+
+from query.database import database_connection
+from query.models.product import Source
+from query.mongodb import find_products
+from query.tables.loaded_tag import append_loaded_tags
+from query.tables.product import (
+    create_minimal_product,
+    delete_products,
+    get_minimal_product,
+    product_fields,
+    update_products_from_staging,
+)
+from query.tables.product_country import fixup_product_countries
+from query.tables.product_ingredient import create_ingredients_from_staging
+from query.tables.product_tags import TAG_TABLES, create_tags_from_staging
+from query.tables.settings import get_last_updated, set_last_updated
+
+logger = logging.getLogger(__name__)
+min_datetime = datetime(1, 1, 1, tzinfo=timezone.utc)
+batch_size = 10000
+
+
+async def get_process_id(connection):
+    return await connection.fetchval("SELECT pg_current_xact_id()")
+
+
+def int_or_none(value):
+    return None if value == None else int(value)
+
+
+async def import_with_filter(filter: Dict, source: Source) -> datetime:
```

Comment:
```suggestion
async def import_with_filter(filter: Dict, source: Source) -> datetime:
    """import/update a set of products (given by filter) using mongodb
    
    :param filter: a mongodb like filter.
      In case the filter is a query on codes,
      it enables removing products corresponding to codes that were not found.
    :param source: depending on the import mechanism, behaviour might change
```